### PR TITLE
Support setting URI SANs in CSRs

### PR DESCRIPTION
**Pull request by eandre**
In #70 support was added to the Vault issuer to allow setting URI SANs.
The particular fix leaves two things to be desired:

(1) It is Vault-specific
(2) It interacts with other configuration options in an unfortunate way.

While (1) is self-explanatory, (2) deserves more explanation.

With Vault, you have a binary knob: either take SANs from the
CSR or from the Vault API request. In order to set URI SANs
with the support in #70, one can configure Vault to take all SANs
from the API request. But if you do this you lose the ability to set
DNS SANs, since this functionality is only supported in Certify through
the base certify.CertConfig which ends up setting SANs on the CSR itself
(which is now being ignored by Vault). If you do not set this flag in
Vault the URI SANs specified through the vault.Issuer are ignored.

To summarize, there is no way with any combination of Certify
configuration and Vault configuration to specify both URI SANs
and DNS SANs.

This commit addresses that limitation, by supporting setting URI SANs
directly on the CSR. This has been supported since Go 1.10.

### CODE REVIEW

**Comment by eandre** [comment]
Test failures seem unrelated to me but I'm not sure what's causing them.

**Comment by eandre** [comment]
Scratch that, I found the issue and it was related :) tests pass now.

File: certify_test.go [hunk 1]
```diff
@@ -146,16 +146,18 @@ var _ = Describe("Certify", func() {
 			case 1:
 				// First call is GetCertificate
 				Expect(in3).To(PointTo(MatchAllFields(Fields{
-					"SubjectAlternativeNames":   Equal(append(cli.CertConfig.SubjectAlternativeNames, serverName, cli.CommonName)),
-					"IPSubjectAlternativeNames": Equal(cli.CertConfig.IPSubjectAlternativeNames),
-					"KeyGenerator":              Not(BeNil()),
+					"SubjectAlternativeNames":    Equal(append(cli.CertConfig.SubjectAlternativeNames, serverName, cli.CommonName)),
+					"IPSubjectAlternativeNames":  Equal(cli.CertConfig.IPSubjectAlternativeNames),
+					"URISubjectAlternativeNames": Equal(cli.CertConfig.URISubjectAlternativeNames),
```
Comment (johanbrandhorst) [inline certify_test.go]: Could you also add some URI's to the `CertConfig` that runs this test, so we get an end-to-end test of this?

**Comment by eandre** [comment]
> This is fantastic, thanks a lot for the contribution! One question; where does this leave Vault URI SANS support? I wouldn't want to remove it, since that would require a major version bump, but it would be nice to add a deprecation warning and encourage the use of this setting instead. What do you think?

They are not _exactly_ identical, since the Vault configuration ends up in the API request instead of in the CSR. If you need to have Vault configured to only read the API request parameters the existing setting is the right place to do that. That said, I'm not sure on exactly what the use case might be for requiring that. It's up to you if you want to add a deprecation notice, since they probably overlap in most use cases (but not all).

File: certify_test.go (hunk 1 above)
Comment (eandre) [reply certify_test.go]: Done

**Comment by johanbrandhorst** [comment]
> They are not _exactly_ identical, since the Vault configuration ends up in the API request instead of in the CSR. If you need to have Vault configured to only read the API request parameters the existing setting is the right place to do that. That said, I'm not sure on exactly what the use case might be for requiring that. It's up to you if you want to add a deprecation notice, since they probably overlap in most use cases (but not all).

Yeah this sounds more likely to confuse users than anything else, I wasn't aware the API parameters would overwrite the CSR. Let's do it. Could you add it to this PR please so users can see the context of the deprecation? Something like

> Warning: it is recommended to use the URISubjectAlternativeNames setting on the
CertConfig struct directly rather than via this Vault specific configuration. Use of this
parameter will override settings on the CSR.

right here: https://github.com/johanbrandhorst/certify/blob/master/issuers/vault/vault.go#L44

This does on some level feel like a Vault bug as well though. I might raise an issue with Vault that URI SANS over the API can override those in the CSR fine but why are they overriding the DNS SANS too?

**Comment by johanbrandhorst** [comment]
Also, presumably this applies to OtherSubjectAlternativeNames too?

**Comment by eandre** [comment]
> Could you add it to this PR please so users can see the context of the deprecation? 
Sure.

> This does on some level feel like a Vault bug as well though. I might raise an issue with Vault that URI SANS over the API can override those in the CSR fine but why are they overriding the DNS SANS too?

The relevant configuration is here: https://www.vaultproject.io/api/secret/pki/index.html#use_csr_sans

It's basically a boolean indicating whether all SANs come from the API or from the CSR. You don't get to choose "some from one and some from the other".

**Comment by eandre** [comment]
I've added documentation to the Vault issuer. Your proposed wording didn't quite match the semantics I've observed from testing with Vault so I came up with a slightly different phrasing.

**Comment by johanbrandhorst** [comment]
> The relevant configuration is here: https://www.vaultproject.io/api/secret/pki/index.html#use_csr_sans
> 
> It's basically a boolean indicating whether all SANs come from the API or from the CSR. You don't get to choose "some from one and some from the other".

We're not setting this, and it's defaults to `true`, so unless the documentation omits this being turned into `false` if you supply some URI sans, I don't think I understand. Is it correct that our current requests should always take the SANS from the CSR? E.g. the `URISubjectAlternativeName` and `OtherSubjectAlternativeName` fields on the Vault configuration should never work?

**Comment by johanbrandhorst** [comment]
> We're not setting this, and it's defaults to `true`, so unless the documentation omits this being turned into `false` if you supply some URI sans, I don't think I understand. Is it correct that our current requests should always take the SANS from the CSR? E.g. the `URISubjectAlternativeName` and `OtherSubjectAlternativeName` fields on the Vault configuration should never work?

Your comments on the vault fields basically answers this question, thank you! I imagine we could still expose the `use_csr_sans` option via the Vault configuration too?

File: issuers/vault/vault.go [hunk 2]
```diff
@@ -41,10 +41,20 @@ type Issuer struct {
 
 	// URISubjectAlternativeNames defines custom URI SANs.
 	// The format is a URI and must match the value specified in allowed_uri_sans, eg spiffe://hostname/foobar
+	//
+	// Warning: By default Vault reads URI SANs directly from the
+	// Certificate Signing Request (CSR), and ignores this field completely.
+	// This field only takes effect when the Vault role has set use_csr_sans to false.
```
Comment (johanbrandhorst) [inline issuers/vault/vault.go]: Can we also say:
```suggestion
	// This field only takes effect when the Vault role has set use_csr_sans to false,
	// and using this setting will ignore any SANs in the CSR.
```

**Comment by eandre** [comment]
No, that vault configuration is a property of the role which must be created ahead of time. It's essentially the policy handed down to the service that is running Certify, not something that Certify itself can conntrol.

File: issuers/vault/vault.go [hunk 3]
```diff
@@ -41,10 +41,20 @@ type Issuer struct {
...
+	//
+	// To configure URI SANs directly in the CSR, set CertConfig.URISubjectAlternativeNames.
 	URISubjectAlternativeNames []string
 
 	// OtherSubjectAlternativeNames defines custom OID/UTF8-string SANs.
 	// The format is the same as OpenSSL: <oid>;<type>:<value> where the only current valid <type> is UTF8.
+	//
+	// Warning: By default Vault reads SANs directly from the
+	// Certificate Signing Request (CSR), and ignores this field completely.
+	// This field only takes effect when the Vault role has set use_csr_sans to false.
```
Comment (johanbrandhorst) [inline issuers/vault/vault.go]:
```suggestion
	// This field only takes effect when the Vault role has set use_csr_sans to false,
	// and using this setting will ignore any SANs in the CSR.
```

**Comment by johanbrandhorst** [comment]
Thanks for your carefully curated git commits by the way, I appreciate it :).

**Comment by eandre** [comment]
> Thanks again, I think I understand now. To summarize, the user should use the `CertConfig` to configure URI SANs, and if they want to configure Other SANs, they have to set the Vault role to have `user_csr_sans` to `false` _and_ they can't use any of the SANs in the CertConfig.
> 
> Does that sound right?

Yes, that sounds right to me.

**Comment by johanbrandhorst** [comment]
Thanks for the contribution, I will prepare another release.

**Comment by eandre** [comment]
Great, thank you!
